voidcrew/modules/shuttle/engine/fuel.dm

```diff
@@ -10,42 +10,82 @@
 
 	icon_state_off = "burst_plasma_off"
 
-	///The specific gas to burn out of the engine heater. If none, burns any gas.
+	///The specific gas to burn out of the engine heater.
 	var/datum/gas/fuel_type
 	///How much fuel (in mols) of the specified gas should be used in a full burn.
 	var/fuel_use = 0
 
+	var/obj/machinery/atmospherics/fueled_engine_heater/connected_heater
+
+/obj/machinery/power/shuttle_engine/ship/fueled/proc/try_link_heater()
+	SHOULD_NOT_OVERRIDE(TRUE)
+	if(connected_heater)
+		return
+
+	var/turf/candidate_turf = get_step(src, dir)
+	var/obj/machinery/atmospherics/fueled_engine_heater/candidate_heater = locate() in candidate_turf
+	if(!candidate_heater || candidate_heater.dir != turn(dir, 180))
+		return
+	on_heater_link(candidate_heater)
+
+/// Called when the engine is linked to a heater; must call parent first.
+/obj/machinery/power/shuttle_engine/ship/fueled/proc/on_heater_link(obj/machinery/atmospherics/fueled_engine_heater/connecting)
+	SHOULD_CALL_PARENT(TRUE)
+	if(connected_heater == connecting)
+		return
+	if(connected_heater)
+		on_heater_unlink()
+
+	connected_heater = connecting
+	connected_heater.connected_engine = src
+
+/// Called when the engine is unlinked from a heater; must call parent first.
+/obj/machinery/power/shuttle_engine/ship/fueled/proc/on_heater_unlink()
+	SHOULD_CALL_PARENT(TRUE)
+	if(!connected_heater)
+		return
+
+	connected_heater.connected_engine = null
+	connected_heater = null
+
+/obj/machinery/power/shuttle_engine/ship/fueled/Initialize()
+	if(!fuel_type)
+		qdel(src)
+		CRASH("Fueled engine [type] did not set fuel_type!")
+	. = ..()
+
+/obj/machinery/power/shuttle_engine/ship/fueled/LateInitialize()
+	. = ..()
+	try_link_heater()
```

Comment:
I didn't originally add this because I wanted the gas to go straight to the engine, but 🤷